### PR TITLE
Fixed parser of OpenCL 3.0 for NVIDIA GPU

### PR DESCRIPTION
**Pull request by redradist**

### CODE REVIEW

**Review by densamoilov — COMMENTED** [review]
Can you please elaborate on the reasoning behind this PR?

**Comment by vpirogov** [comment]
@redradist,

Adding 3.0 to the list of OpenCL version makes sense. We are still looking for your comments on the rationale for remaining changes.

**Comment by redradist** [comment]
@vpirogov @densamoilov Because this code successfully detect version of OpenCL library, but this one with CMAKE_CHECK_SYMBOLS cannot, even with defined variables to include dirs and link libraries:
```cmake
...

function(_FIND_OPENCL_VERSION)
  include(CheckSymbolExists)
  include(CMakePushCheckState)
  set(CMAKE_REQUIRED_QUIET ${OpenCL_FIND_QUIETLY})
  set(CMAKE_REQUIRED_INCLUDES ${OpenCL_INCLUDE_DIR})
  set(CMAKE_REQUIRED_LIBRARIES ${OpenCL_LIBRARY})

  CMAKE_PUSH_CHECK_STATE()
  foreach(VERSION "3_0" "2_2" "2_1" "2_0" "1_2" "1_1" "1_0")
    if(APPLE)
      CHECK_SYMBOL_EXISTS(
        CL_VERSION_${VERSION}
        "${OpenCL_INCLUDE_DIR}/Headers/cl.h"
        OPENCL_VERSION_${VERSION})
    else()
      CHECK_SYMBOL_EXISTS(
        CL_VERSION_${VERSION}
        "${OpenCL_INCLUDE_DIR}/CL/cl.h"
        OPENCL_VERSION_${VERSION})
    endif()

    message(STATUS "OPENCL_VERSION_${VERSION} = ${OPENCL_VERSION_${VERSION}}")

    if(OPENCL_VERSION_${VERSION})
      string(REPLACE "_" "." VERSION "${VERSION}")
      set(OpenCL_VERSION_STRING ${VERSION} PARENT_SCOPE)
      string(REGEX MATCHALL "[0-9]+" version_components "${VERSION}")
      list(GET version_components 0 major_version)
      list(GET version_components 1 minor_version)
      set(OpenCL_VERSION_MAJOR ${major_version} PARENT_SCOPE)
      set(OpenCL_VERSION_MINOR ${minor_version} PARENT_SCOPE)
      break()
    endif()
  endforeach()
  CMAKE_POP_CHECK_STATE()
endfunction()

...
```

`OPENCL_VERSION_${VERSION}` is always empty, due to possible missed defines

File: cmake/FindOpenCL.cmake [hunk 1]
```diff
@@ -42,26 +42,56 @@ if(opencl_root_hints)
...
-  CMAKE_PUSH_CHECK_STATE()
-  foreach(VERSION "2_2" "2_1" "2_0" "1_2" "1_1" "1_0")
-    set(CMAKE_REQUIRED_INCLUDES "${OpenCL_INCLUDE_DIR}")
-
-    if(APPLE)
-      CHECK_SYMBOL_EXISTS(
-        CL_VERSION_${VERSION}
-        "${OpenCL_INCLUDE_DIR}/Headers/cl.h"
-        OPENCL_VERSION_${VERSION})
+  foreach(VERSION "3_0" "2_2" "2_1" "2_0" "1_2" "1_1" "1_0")
```
Comment (densamoilov) [inline cmake/FindOpenCL.cmake:45]: (random place)

There are a few things that look suspicious to me:
* We didn't implement FindOpenCL.cmake module, we've taken it from CMake (see [here](https://github.com/Kitware/CMake/blob/master/Modules/FindOpenCL.cmake)) and it still uses `CHECK_SYMBOL_EXISTS` and the whole community seems to be fine with it
* The version macros are standardized by the OpenCL specification so if the macros are missing then it likely means that there is something wrong with the OpenCL implementation
* It should be fine to check `cl.h` header as it includes `cl_version.h` header and `CHECK_SYMBOL_EXISTS` does perform a preprocessing 

Given all this, I think the issue you encounter is somewhere on your end.

Comment (redradist) [reply cmake/FindOpenCL.cmake:45]: @densamoilov I have figured out that `CHECK_SYMBOL_EXISTS` is broken on 3.27.7 version, but it is working on 3.30.2

`cl.h` file I use is completely fine, it is OpenCL version for NVIDIA GPU, see file in attachments below:

[cl_files.zip](https://github.com/user-attachments/files/16529025/cl_files.zip)

Comment (redradist) [reply cmake/FindOpenCL.cmake:45]: Here is my OpenCL files for NVIDIA GPU:
[cl_files.zip](https://github.com/user-attachments/files/16529151/cl_files.zip)

**Comment by vpirogov** [comment]
Considering the discussion the best way forward is updating FindOpenCL.cmake from upstream.

**Comment by redradist** [comment]
@densamoilov @vpirogov The CMake version says that it supports [(Detection of OpenCL 2.1 and 2.2)](https://github.com/Kitware/CMake/blob/master/Modules/FindOpenCL.cmake#L13), also I've found that already someone implemented the similar version of FindOpenCL by trying to compile example project.
I think that approach is more stable

**Comment by vpirogov** [comment]
@redradist, if [FindOpenCL.cmake](https://github.com/Kitware/CMake/blob/master/Modules/FindOpenCL.cmake) from upstream CMake works in your scenario please update oneDNN copy accordingly. Be mindful of [SYCL related changes](https://github.com/oneapi-src/oneDNN/commit/e73665ff86929cdbd2865a8d2432fb9ee71b7335#diff-d527a18f665b4324fa4edb61db59776f6c57aa12dda532c583ca94fbdd67e3b5) that are local to oneDNN copy though.

**Comment by redradist** [comment]
> @redradist, if [FindOpenCL.cmake](https://github.com/Kitware/CMake/blob/master/Modules/FindOpenCL.cmake) from upstream CMake works in your scenario please update oneDNN copy accordingly. Be mindful of [SYCL related changes](https://github.com/oneapi-src/oneDNN/commit/e73665ff86929cdbd2865a8d2432fb9ee71b7335#diff-d527a18f665b4324fa4edb61db59776f6c57aa12dda532c583ca94fbdd67e3b5) that are local to oneDNN copy though.

@vpirogov 
No, it does not work. It supports only OpenCL 2.2 as I mentioned.
Test project compilation looks for me more reliable

**Comment by vpirogov** [comment]
> No, it does not work. It supports only OpenCL 2.2 as I mentioned.

Is this conclusion based on the comment in the file or actual testing? I see the comment comment that says 'Detection of OpenCL 2.1 and 2.2.', however commit history for the file includes patch indicating 3.0 support: https://github.com/Kitware/CMake/commit/587d1942b56b22628d5f03aeb41819efe21b68a4.

We would like to avoid creating and maintaining completely new FindOpenCL.cmake.

**Comment by redradist** [comment]
> > No, it does not work. It supports only OpenCL 2.2 as I mentioned.
> 
> Is this conclusion based on the comment in the file or actual testing?

It is based on actual testing. You could verify it by yourself, just install OpenCL driver for NVIDIA GPU on Ubuntu and you'll see what the problem.
My solution resolve issue completly

**Comment by redradist** [comment]
> > No, it does not work. It supports only OpenCL 2.2 as I mentioned.
> 
> Is this conclusion based on the comment in the file or actual testing? I see the comment comment that says 'Detection of OpenCL 2.1 and 2.2.', however commit history for the file includes patch indicating 3.0 support: [Kitware/CMake@587d194](https://github.com/Kitware/CMake/commit/587d1942b56b22628d5f03aeb41819efe21b68a4).
> 
> We would like to avoid creating and maintaining completely new FindOpenCL.cmake.

@vpirogov One additional comment, previously this `cmake/FindOpenCL.cmake` worked properly, but since NVIDIA update they driver with support of OpenCL 3.0, it is broken

**Comment by vpirogov** [comment]
> One additional comment, previously this `cmake/FindOpenCL.cmake` worked properly, but since NVIDIA update they driver with support of OpenCL 3.0, it is broken

Could you please indicate which NVIDIA OpenCL driver package you are testing with? I'm running Ubuntu 22.04 with the latest NVIDIA driver, but FindOpenCL.cmake from this PR fails:
```
$ apt list --installed | grep opencl

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

cuda-opencl-12-6/unknown,unknown,now 12.6.68-1 amd64 [installed,upgradable to: 12.6.77-1]
cuda-opencl-dev-12-6/unknown,unknown,now 12.6.68-1 amd64 [installed,upgradable to: 12.6.77-1]
ocl-icd-libopencl1/jammy,now 2.2.14-3 amd64 [installed,automatic]
(base) vpirogov@jfldnn043:~/work/dnnl/build$ ls
CheckOpenCLSymbol  CMakeCache.txt  CMakeFiles
(base) vpirogov@jfldnn043:~/work/dnnl/build$ rm -rf *
(base) vpirogov@jfldnn043:~/work/dnnl/build$ cmake .. -DONEDNN_GPU_RUNTIME=OCL
CMake Deprecation Warning at CMakeLists.txt:17 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- CMAKE_BUILD_TYPE is unset, defaulting to Release
-- The C compiler identification is GNU 11.3.0
-- The CXX compiler identification is GNU 11.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- DNNL_TARGET_ARCH: X64
-- DNNL compat: set DNNL_GPU_RUNTIME to ONEDNN_GPU_RUNTIME with value `OCL`
-- DNNL_LIBRARY_NAME: dnnl
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Found OpenMP_C: -fopenmp (found version "4.5")
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Found OpenMP: TRUE (found version "4.5")
-- GPU support is enabled (OpenCL)
-- Found OpenCL headers: OpenCL_INCLUDE_DIR-NOTFOUND
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
/nfs/pdx/home/vpirogov/work/dnnl/build/CheckOpenCLSymbol/OpenCL_INCLUDE_DIR
   used as include directory in directory /nfs/pdx/home/vpirogov/work/dnnl/build/CheckOpenCLSymbol
OpenCL_LIBRARY
    linked by target "TestOpenCL" in directory /nfs/pdx/home/vpirogov/work/dnnl/build/CheckOpenCLSymbol

CMake Error at cmake/FindOpenCL.cmake:78 (try_compile):
  Failed to generate test project build system.
Call Stack (most recent call first):
  cmake/FindOpenCL.cmake:189 (_FIND_OPENCL_VERSION)
  cmake/OpenCL.cmake:35 (find_package)
  CMakeLists.txt:130 (include)


-- Configuring incomplete, errors occurred!
```

Environment details:
```
$ uname -a
Linux jfldnn043 5.15.0-56-generic #62-Ubuntu SMP Tue Nov 22 19:54:14 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

$ apt list --installed | grep nvidia-driver

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

nvidia-driver-560-open/unknown,now 560.35.03-0ubuntu1 amd64 [installed,upgradable to: 560.35.05-0ubuntu1]

$ apt list --installed | grep opencl

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

cuda-opencl-12-6/unknown,unknown,now 12.6.68-1 amd64 [installed,upgradable to: 12.6.77-1]
cuda-opencl-dev-12-6/unknown,unknown,now 12.6.68-1 amd64 [installed,upgradable to: 12.6.77-1]
ocl-icd-libopencl1/jammy,now 2.2.14-3 amd64 [installed,automatic]
```

FindOpenCL.cmake from upstream works fine in this environment.

**Comment by vpirogov** [comment]
> I'm using the following OpenCL packages:

This is Kronos standard OpenCL loader and headers. It's vendor neutral and works perfectly fine with current FindOpenCL.cmake as tested on clean Ubuntu 22.04 image. This PR fails though:
```

```

Looks like the problem you are trying to address is somewhere else.

**Comment by redradist** [comment]
@vpirogov Issue was more deeply investigated and new PR was opened with smaller fix, see https://github.com/oneapi-src/oneDNN/pull/2378

**Pull request closed without merging**